controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -803,6 +803,15 @@ private long findSegmentSplitsMerges(List<Segment> referenceSegmentsList, List<S
                                 historyTable.getData()))));
     }
 
+    @Override
+    public CompletableFuture<UUID> generateNewTxnId(int msb32Bit, long lsb64Bit) {
+        return getLatestEpoch()
+                .thenApply(epoch -> {
+                    long msb64Bit = (long) epoch.getKey() << 32 | msb32Bit & 0xFFFFFFFFL;
+                    return new UUID(msb64Bit, lsb64Bit);
```

Comment:
Shouldn't this be a `BigLong`? I'd much rather pass around those kinds of objects than UUIDs.

Reply:
Transation ids are used as UUID everywhere in the code. Hence we have not changed that and as part of this change we compose this UUID based on some scheme so that its not random and captures additional information wrt epoch apart from being a unique id.

Reply:
I think that @shiveshr is saying that once he completes the set of changes for rolling txns, this method will return `BigLong`, is it right @shiveshr ?

Reply:
I would still like to know the answer to this.

Reply:
I have given details of this here:
https://github.com/pravega/pravega/pull/2517#issuecomment-392575330

Brief summary, no. Transaction ids continue to be UUIDs. thats how controller and clients use it. 

Segment store cares about fully qualified names only. 

Controller ensures that it creates top level segments and their duplicates such that they are mapped to same segment containers. 
And it would either merge txn segment into original or one of the duplicates that it created (in case of rolling transaction). 

So transaction id continues to be UUID. We use biglong to generate txn Id GUID that follows a predefined scheme where we can extract `epoch` from txn id so that we can go to the correct group of txns and consolidate multiple merges together.